src/main/java/at/ac/tuwien/kr/alpha/common/rule/AbstractRule.java

```diff
@@ -0,0 +1,120 @@
+package at.ac.tuwien.kr.alpha.common.rule;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+
+import at.ac.tuwien.kr.alpha.Util;
+import at.ac.tuwien.kr.alpha.common.atoms.Literal;
+import at.ac.tuwien.kr.alpha.common.rule.head.Head;
+
+/**
+ * An abstract representation of a rule with a specific type of @{link Head} (type parameter H)
+ * 
+ * @param <H> the type of head for this rule Copyright (c) 2019, the Alpha Team.
+ * 
+ *        Copyright (c) 2017-2019, the Alpha Team.
+ */
+public abstract class AbstractRule<H extends Head> {
+
+	private final H head;
+	private final List<Literal> body;
+
+	public AbstractRule(H head, List<Literal> body) {
+		this.head = head;
+		// Remove duplicate body literals.
+		LinkedHashSet<Literal> bodyLiterals = new LinkedHashSet<>(body);
+		this.body = new ArrayList<>(bodyLiterals);
+
+		if (!this.isSafe()) {
+			// TODO: safety check needs to be adapted to solver what the solver actually understands. Will change in the future, adapt exception message
+			// accordingly.
+			throw new RuntimeException("Encountered unsafe rule: " + toString() + System.lineSeparator()
+					+ "Notice: A rule is considered safe if all variables occurring in negative literals, builtin atoms, and the head of the rule also occur in some positive literal.");
+		}
```

Comment:
I'm not so sure about that.. The way I understand it, **every** rule needs to be safe.
While it may not be possible to do this check on the abstract level, it still holds that the check must be performed for every (type of) rule. To reflect that, I added a note to the current dummy-implementation of `isSafe` that concrete implementations should probably be in subclasses and `isSafe` should be `protected abstract` in `AbstractRule`.
However, I`d like to keep the check for `isSafe` in the `AbstractRule` constructor, since that ensures that it's not possible for implementations to skip the safety check.